### PR TITLE
mongo-cxx-driver: use validate() for checking configuration

### PR DESCRIPTION
**Pull request by mathbunnyru**
Specify library name and version:  **mongo-cxx-driver/3.6.2**

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/mongo-cxx-driver/all/conanfile.py [hunk 1]
```diff
@@ -148,15 +123,41 @@ def _configure_cmake(self):
         self._cmake.configure(build_folder=self._build_subfolder)
         return self._cmake
 
+    def validate(self):
+        if self.options.with_ssl and not bool(self.options["mongo-c-driver"].with_ssl):
+            raise ConanInvalidConfiguration("mongo-cxx-driver with_ssl=True requires mongo-c-driver with a ssl implementation")
```
Comment (mathbunnyru) [inline recipes/mongo-cxx-driver/all/conanfile.py:128]: No new code here, everything is from `configure()` or `build()` methods.

I followed the recommendation from here:
https://blog.conan.io/2020/12/14/New-conan-release-1-32.html
the `validate()` method should take over all responsibility for raising `ConanInvalidConfiguration` errors

**Comment by mathbunnyru** [comment]
`conan create all mongo-cxx-driver/3.6.2@ --build=missing -o mongo-cxx-driver:with_ssl=True -o mongo-c-driver:with_ssl=False` now doesn't build dependencies, which is awesome.

**Comment by prince-chrismc** [comment]
Should the be a new hook?

File: recipes/mongo-cxx-driver/all/conanfile.py [hunk 2]
```diff
@@ -148,15 +123,41 @@ def _configure_cmake(self):
...
+            raise ConanInvalidConfiguration("experimental polyfill is not supported for apple-clang")
+
+        if compiler not in self._compilers_minimum_version:
+            self.output.warn("Unknown compiler, assuming it supports at least C++{}".format(self._minimal_std_version))
+            return
+
+        version = tools.Version(self.settings.compiler.version)
+        if version < self._compilers_minimum_version[compiler]:
+            raise ConanInvalidConfiguration(
+                "{} requires a compiler that supports at least C++{}".format(
```
Comment (prince-chrismc) [inline recipes/mongo-cxx-driver/all/conanfile.py:149]: I'll have to read that blog post but I would have expected this in configure method

**Comment by SpaceIm** [comment]
I would move `validate()` between `requirements()` and `source()`.

**Comment by mathbunnyru** [comment]
> Should the be a new hook?

For now I created the issue https://github.com/conan-io/conan/issues/8278

**Comment by mathbunnyru** [comment]
> I would move `validate()` between `requirements()` and `source()`. It helps to understand the logic and eventually consequences.
> 
> You can also bump mongo-c-driver to 1.17.3

Done. Also updated boost.

**Review by prince-chrismc — DISMISSED** [review]
very weak approval, I dont know this new method 🙊

**Review by Croydon — APPROVED** [review]

**Comment by mathbunnyru** [comment]
Both errors are: `ERROR: Missing prebuilt package for 'mongo-c-driver/1.17.3'`
What should I do to fix this?

**Comment by SpaceIm** [comment]
Indeed:
- with `conan search mongo-c-driver/1.17.3@ --table=file.html -r conan-center`, I see 87 packages (1 is missing)
- 1.17.3 is not even displayed on https://conan.io/center/mongo-c-driver

**Comment by prince-chrismc** [comment]
> What should I do to fix this?

Call the ~ghostbusters~ infra team!

team

//cc @jgsogo @danimtb if you guys have some time to take a peek at the issue we always appreciate it!

**Comment by mathbunnyru** [comment]
_If it's somethin' weird an' it don't look good
Who ya gonna call?
(Ghostbusters!)_

https://github.com/conan-io/conan-center-index/issues/4076

**Comment by mathbunnyru** [comment]
I decided to stick with the current version of mongo-c-driver, to merge this PR faster.
The version could be updated later without any problems.

**Comment by prince-chrismc** [comment]
🤞 CI passes... dang missing binaries

**Comment by mathbunnyru** [comment]
Merged master, there was an important fix to this recipe.

**Comment by mathbunnyru** [comment]
> 🤞 CI passes... dang missing binaries

Let's hope and pray :)

**Review by prince-chrismc — APPROVED** [review]

**Review by SpaceIm — APPROVED** [review]

**Comment by mathbunnyru** [comment]
@Croydon may I ask you to review again, please?

**Review by SSE4 — APPROVED** [review]

**Pull request merged**